File: mac/KhaytCore/Sources/KhaytCore/JS/zatca-submit.js
'use strict';
(function (global) {

/**
 * Reporting a Saudi tax invoice to ZATCA, Phase 2.
 *
 * Wrapped like `lib/zatca-qr.js` beside it, and for the same reason that file
 * gives: these rules lived in `renderer/invoicing.js`, so the only thing that
 * could decide whether an invoice had been accepted was the Electron window.
 * The module existed and had a test suite; `renderer/invoicing.js` kept its own
 * copy of four of these functions and it was the copy that ran. They agreed —
 * checked line by line before this was written — but that was luck holding,
 * not a guarantee, and the tests were on the half nobody executed.
 *
 * What these decide is not cosmetic. `nextZatcaIcv` numbers an invoice in a
 * sequence a tax authority requires to be unbroken, and `zatcaSubmitAccepted`
 * decides whether a document handed to a customer counts as reported.
 */

/**
 * Base64-encode UTF-8 XML for the ZATCA API payload.
 *
 * `Buffer` in the main process, `btoa` in a window — and neither exists in the
 * other. The caller may pass its own encoder; the fallbacks cover both hosts so
 * that most callers do not have to care. `zatca-qr.js` solves the same split by
 * taking `base64` as an option, and this keeps that shape.
 */
function xmlToBase64(xml, opts) {
  const text = String(xml || '');
  if (opts && typeof opts.base64 === 'function') return opts.base64(text);
  if (typeof Buffer !== 'undefined') return Buffer.from(text, 'utf8').toString('base64');
  const bytes = new TextEncoder().encode(text);
  let binary = '';
  for (let i = 0; i < bytes.length; i++) binary += String.fromCharCode(bytes[i]);
  return btoa(binary);
}

/** Whether Phase 2 submission prerequisites are met. */
function zatcaPhase2Ready(settings) {
  const z2 = settings?.zatcaPhase2;
  return !!(settings?.enableZatca && z2?.enabled && (z2.pcsid || z2.csid));
}

/** Next invoice counter value for a new submission (reuse pending ICV on retry). */
function nextZatcaIcv(z2, order) {
  if (order?.zatcaSubmission?.icv) return order.zatcaSubmission.icv;
  return (z2?.invoiceCounter || 0) + 1;
}

/** True when order is eligible for ZATCA reporting. */
function orderEligibleForZatcaSubmit(order) {
  if (!order || order.voidedAt) return false;
  return order.status === 'completed' || order.status === 'delivered';
}

/** Interpret Fatoorah HTTP response as accepted/rejected. */
function zatcaSubmitAccepted(httpOk, body) {
  if (!httpOk) return false;
  const status = body?.validationResults?.status || body?.reportingStatus || body?.clearanceStatus;
  if (status && String(status).toUpperCase() === 'REJECTED') return false;
  return true;
}

/** Build submission log entry (trimmed for store). */
function buildZatcaLogEntry({ order, payload, result, manual, status, message }) {
  return {
    orderId: order.id,
    invoiceNumber: payload.invoiceNumber,
    uuid: payload.uuid,
    icv: payload.invoiceCounter,
    at: new Date().toISOString(),
    status,
    httpStatus: result?.status ?? null,
    message: message || '',
    manual: !!manual,
  };
}

function appendZatcaSubmissionLog(z2, entry) {
  const log = Array.isArray(z2.submissions) ? [...z2.submissions] : [];
  log.unshift(entry);
  z2.submissions = log.slice(0, 100);
}

const api = {
  xmlToBase64,
  zatcaPhase2Ready,
  nextZatcaIcv,
  orderEligibleForZatcaSubmit,
  zatcaSubmitAccepted,
  buildZatcaLogEntry,
  appendZatcaSubmissionLog,
};
if (typeof module !== 'undefined' && module.exports) module.exports = api;
global.KhaytZatcaSubmit = api;
})(typeof globalThis !== 'undefined' ? globalThis : this);

File: mac/KhaytCore/Sources/KhaytCore/ZatcaSubmit.swift
import Foundation

/// Reporting a Saudi tax invoice to ZATCA, Phase 2.
///
/// What these decide is not cosmetic. The invoice counter numbers a document in
/// a sequence a tax authority requires to be UNBROKEN, and `accepted` decides
/// whether an invoice already handed to a customer counts as reported.
public enum ZatcaSubmit {

    /// Base64-encode UTF-8 XML for the API payload.
    ///
    /// `Buffer` in a main process, `btoa` in a window — and neither exists in
    /// the other, which is why the original takes an encoder. Here there is
    /// one encoder and no split to paper over.
    public static func xmlToBase64(_ xml: String) -> String {
        Data(xml.utf8).base64EncodedString()
    }

    /// Whether Phase 2 submission prerequisites are met.
    public static func phase2Ready(settings: [String: JSONValue]) -> Bool {
        guard JSSemantics.truthy(settings["enableZatca"]),
              case .object(let z2)? = settings["zatcaPhase2"],
              JSSemantics.truthy(z2["enabled"]) else { return false }
        return JSSemantics.truthy(z2["pcsid"]) || JSSemantics.truthy(z2["csid"])
    }

    /// The counter value for a new submission — or the pending one again, so a
    /// retry does not burn a number out of a sequence that must not have gaps.
    ///
    /// ── THE ORIGINAL CONCATENATES, AND THIS REPRODUCES IT ─────────────────
    ///
    /// `(z2?.invoiceCounter || 0) + 1`. A counter stored as the STRING "5"
    /// is truthy, so `+ 1` appends rather than adds and the next invoice is
    /// numbered 51. The book writes a number and no shop has hit this, but a
    /// port that quietly answered 6 would be two apps disagreeing about a
    /// sequence a tax authority checks — so the fault is carried over, in the
    /// open, with a test on it, rather than fixed on one side only.
    public static func nextIcv(z2: JSONValue?, order: JSONValue?) -> JSONValue {
        if case .object(let o)? = order, case .object(let sub)? = o["zatcaSubmission"],
           let icv = sub["icv"], JSSemantics.truthy(icv) {
            return icv
        }
        var counter: JSONValue = .number(0)
        if case .object(let z)? = z2, let c = z["invoiceCounter"], JSSemantics.truthy(c) {
            counter = c
        }
        // `counter + 1` is JavaScript's `+`, which is addition only when both
        // sides are primitives that are not strings. An array or an object is
        // coerced to text first, so `[] + 1` is "1" and `{} + 1` is
        // "[object Object]1" — the harness found both.
        switch counter {
        case .number(let n): return .number(n + 1)
        case .bool(let b): return .number((b ? 1 : 0) + 1)
        default: return .string(JSSemantics.text(counter) + "1")
        }
    }

    /// True when an order is eligible for ZATCA reporting. A job still on the
    /// bench is not late to be reported.
    public static func eligible(_ order: JSONValue?) -> Bool {
        guard case .object(let o)? = order, !JSSemantics.truthy(o["voidedAt"]) else { return false }
        guard case .string(let status)? = o["status"] else { return false }
        return status == "completed" || status == "delivered"
    }

    /// Read a Fatoorah response as accepted or rejected.
    ///
    /// Anything that is not an explicit REJECTED counts as accepted, which is
    /// the original's shape: the authority answers with a status under one of
    /// three names depending on the endpoint, and a response that names none of
    /// them is not a rejection.
    public static func accepted(httpOk: Bool, body: JSONValue?) -> Bool {
        guard httpOk else { return false }
        var fields: [String: JSONValue] = [:]
        if case .object(let o)? = body { fields = o }
        var status: JSONValue?
        if case .object(let results)? = fields["validationResults"],
           let s = results["status"], JSSemantics.truthy(s) {
            status = s
        } else if let s = fields["reportingStatus"], JSSemantics.truthy(s) {
            status = s
        } else if let s = fields["clearanceStatus"], JSSemantics.truthy(s) {
            status = s
        }
        guard let status, JSSemantics.truthy(status) else { return true }
        // `.toUpperCase()`, which is locale-independent on both sides.
        return JSSemantics.text(status).uppercased() != "REJECTED"
    }

    /// One line of the submission log, trimmed for the store.
    ///
    /// `at` is passed in rather than read from a clock: the original calls
    /// `new Date()` inside, which is the one thing in the module that stops it
    /// being testable.
    public static func logEntry(order: JSONValue, payload: JSONValue, httpStatus: Int?,
                                manual: Bool, status: String, message: String,
                                at: String) -> JSONValue {
        var o: [String: JSONValue] = [:]
        if case .object(let fields) = order { o = fields }
        var p: [String: JSONValue] = [:]
        if case .object(let fields) = payload { p = fields }
        return .object([
            "orderId": o["id"] ?? .null,
            "invoiceNumber": p["invoiceNumber"] ?? .null,
            "uuid": p["uuid"] ?? .null,
            "icv": p["invoiceCounter"] ?? .null,
            "at": .string(at),
            "status": .string(status),
            "httpStatus": httpStatus.map { JSONValue.number(Double($0)) } ?? .null,
            "message": .string(message),
            "manual": .bool(manual),
        ])
    }

    /// The submission log with one more line on the front, capped.
    ///
    /// Newest first, and only the last hundred: a log that grows for ever is a
    /// store that grows for ever, and it syncs.
    public static func appendingLog(_ existing: JSONValue?, _ entry: JSONValue) -> [JSONValue] {
        var log: [JSONValue] = []
        if case .array(let rows)? = existing { log = rows }
        log.insert(entry, at: 0)
        return Array(log.prefix(100))
    }
}
